### PR TITLE
Implement two new RAII utilities and migrate nn

### PR DESCRIPTION
**Pull request by jackgerrits**
- `VW::scope_exit` - for exception safe resource releasing based on scopes
- `VW::swap_guard` - scope based swapping of variables for restoring object state

### CODE REVIEW

File: vowpalwabbit/nn.cc [hunk 1]
```diff
@@ -243,7 +230,7 @@ void predict_or_learn_multi(nn& n, single_learner& base, example& ec)
     float save_final_prediction = 0;
     float save_ec_loss = 0;
 
-CONVERSE:  // That's right, I'm using goto.  So sue me.
+  CONVERSE:  // That's right, I'm using goto.  So sue me.
```
Comment (peterychang) [inline vowpalwabbit/nn.cc:233]: IIRC labels need to be at the beginning of the line

Comment (jackgerrits) [reply vowpalwabbit/nn.cc:233]: This was clang-format being "helpful" will revert it

File: vowpalwabbit/nn.cc [hunk 2]
```diff
@@ -169,7 +152,11 @@ void predict_or_learn_multi(nn& n, single_learner& base, example& ec)
   shared_data sd;
   memcpy(&sd, n.all->sd, sizeof(shared_data));
   {
-    sd_guard(n.all, &sd);
+    auto* saved_all = n.all;
+    auto* saved_sd = n.all->sd;
+    // guard for all.sd as it is modified - this will restore the state at the end of the scope.
+    auto sd_guard = VW::scope_guard([saved_all, saved_sd]() { saved_all->sd = saved_sd; });
```
Comment (peterychang) [inline vowpalwabbit/nn.cc:158]: I'm not sure if I like this or if I hate it... On one hand its a generic guard, on the other it still requires a lot of manual setup that leaves a lot of room for mistakes

Comment (jackgerrits) [reply vowpalwabbit/nn.cc:158]: I think this is because the old guard wasn't actually just a guard but it also handed setting up the state of the all->sd pointer to a new value. This separates the concerns to /just/ the cleanup action

Comment (peterychang) [reply vowpalwabbit/nn.cc:158]: Guards typically handle the setup and teardown of the objects being passed in; think of a lock guard. The part I dislike is that you need to manually do the setup while the teardown is automated. It feels like you could get into a lot of trouble if they're not paired together

Comment (jackgerrits) [reply vowpalwabbit/nn.cc:158]: Yeah I guess it is kind of a taste thing. If there is a pattern that enables reuse of a guard then a specific object may make sense - but for one offs it seems excessive to create lots of objects with custom behavior. Renaming it to scope_exit may be better?

This pattern is inspired by the WIL (Windows internal library) scope_exit functionality: https://github.com/microsoft/wil/blob/cd51fa9d3c24e9e9f5d4cdf1a9768d68f441fc48/include/wil/resource.h#L543

File: vowpalwabbit/guard.h [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+  swap_guard_impl(const swap_guard_impl&) = delete;
+  swap_guard_impl& operator=(const swap_guard_impl&) = delete;
+
+  swap_guard_impl(swap_guard_impl&& other) noexcept
+      : _original_location(other._original_location), _value_to_swap(other._value_to_swap), _will_swap_back(other._will_swap_back)
+  {
+    other._will_swap_back = false;
+  }
+
+  swap_guard_impl& operator=(swap_guard_impl&& other) noexcept
```
Comment (peterychang) [inline vowpalwabbit/guard.h:34]: Do we need the assignment operator? Seems safer to delete this too

Comment (peterychang) [reply vowpalwabbit/guard.h:34]: Oh, I see, this is a move assignment.. Still though, do we need it?

Comment (jackgerrits) [reply vowpalwabbit/guard.h:34]: Good point - will remove

File: vowpalwabbit/guard.h [hunk 4]
```diff
@@ -0,0 +1,136 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+#pragma once
+
+#include <utility>
+#include "future_compat.h"
+
+// While it is a general concept this specific implemenation of scope based cleanup was inspired by
+// https://github.com/microsoft/wil/blob/cd51fa9d3c24e9e9f5d4cdf1a9768d68f441fc48/include/wil/resource.h#L543
```
Comment (peterychang) [inline vowpalwabbit/guard.h:10]: This is in the wrong file

File: vowpalwabbit/guard.h [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+ public:
+  swap_guard_impl(T& original_location, T& value_to_swap) noexcept : _original_location(original_location), _value_to_swap(value_to_swap)
+  {
+    std::swap(_original_location, _value_to_swap);
+  }
+
+  swap_guard_impl(const swap_guard_impl&) = delete;
+  swap_guard_impl& operator=(const swap_guard_impl&) = delete;
+
+  swap_guard_impl(swap_guard_impl&& other) noexcept
```
Comment (peterychang) [inline vowpalwabbit/guard.h:28]: the move constructors aren't necessarily safe for all of our objects. If they're not necessary, I'd rather delete them to be safe

Comment (jackgerrits) [reply vowpalwabbit/guard.h:25]: These are definitely needed as they permit the make functions.
